### DISCUSSION TITLE
Trying to run an exe from stlite desktop app

### DISCUSSION THREAD

**Discussion by caleb-depotanalytics** [whitphx/stlite#531] [Q&A]
Hey!

Since I need pyodbc for a Desktop app that I am creating, I am using pyinstaller to create a backend app that will serve as sort of a local API hub (with just cmd line options) for my non-pure python calls.

As a result, from the stlite/desktop app, I need to run a file called "database.exe" which can be any known location on the filesystem. I tried installing it in C:\Program Files\database.exe, and the streamlit desktop app couldn't find it:

Hopefully, this image illustrates the situation well.

![image](https://user-images.githubusercontent.com/104788759/227330093-744451c1-528a-4c19-bc57-5b1ce6f14ee7.png)

Could anyone point me toward the best way to solve this problem?

**Comment by whitphx** [comment]
Unfortunately, it is not possible at this point, as long as I understand the situation correctly.

It is because, in short, Streamlit apps on the `@stlite/desktop` runtime are not allowed to access the host file system by design, and for security reasons.

---
In details, the stlite runtime is "sandboxed".
When you launch a stlite app, first [Pyodide](https://pyodide.org/en/stable/) is started, then Streamlit is run on it. And Pyodide has [its own virtual file system](https://pyodide.org/en/stable/usage/file-system.html) that the Python code running on Pyodide can only have access to.
Then, a Streamlit app running on stlite can only access to this virtual file system, and `@stlite/desktop` controls it so that only the `build` directory's contents are mounted to the virtual FS to be input to the stlite runtime.
So, currently there is no way for Streamlit apps running in `@stlite/desktop` to access anything outside the `build` directory .

---

There is a related issue: https://github.com/whitphx/stlite/issues/463, which may require access to the host file system.
In addition to it, please create an issue as a feature request and continue the discussion on what kind of feature or API stlite should have to fulfill this need.

**Comment by caleb-depotanalytics** [comment]
Could I add the exe to the build directory as an external resource and call it from there?

**Comment by whitphx** [comment]
Probably not.

You can bundle the exe file in the `build` directory, which will be copied to the virtual file system and stlite can read it then. However, the binary which the stlite app can see is a copy on the virtual file system. So it is still not possible for the stlite app to access the exe file on the host file system.

**Comment by whitphx** [comment]
If launching another exe file is the purpose, launching an app with a URL scheme may work, although I haven't tried it.
https://learn.microsoft.com/en-us/windows/uwp/launch-resume/launch-default-app

As the desktop app of `@stlite/desktop` is a kind of web app which is running on an Electron-bundled Chrome, such web-based technique should be able to be used.

**Reply by caleb-depotanalytics** [reply]
Thanks so much for this - I actually got to the answer. If I just place my exe (`database.exe`) in streamlit_app directory. and called `database.exe` directly, likeso:

```python3
if sys.platform == "emscripten":
    db_call_exe = Path("database.exe") # Path("C:\\") / "Program Files" / "database.exe"
    if not db_call_exe.exists():
        st.write(f"database.exe not found in {str(db_call_exe)}")
        st.write(Path(".").glob("*"))
else: # sys.platform == "win32":
    db_call_exe = Path(__file__).parent.parent.parent / "database-utility" / "database.exe"
```

These two exes will be inseparably linked, so this is not an issue at all for me to do.

I'd be interested to try to URL scheme to, but I think this is fine for our case.

**Reply by caleb-depotanalytics** [reply]
Actually, nevermind - didn't realize I was getting some data from the cache earlier and not actually calling the .exe:
```
OSError: [Errno 52] Function not implemented
Traceback:
File "/lib/python3.10/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 564, in _run_script
    exec(code, module.__dict__)
File "/home/pyodide/streamlit_app.py", line 65, in <module>
    add_new_project_template()
File "/home/pyodide/streamlit_app.py", line 24, in add_new_project_template
    projects = get_all_project_types()
File "/home/pyodide/workstream/data_calls.py", line 53, in get_all_project_types
    return exec_db_call("get_all_project_types")
File "/home/pyodide/workstream/data_calls.py", line 42, in exec_db_call
    subprocess.call(call_list, stdout=sys.stdout)
File "/lib/python3.10/subprocess.py", line 345, in call
    with Popen(*popenargs, **kwargs) as p:
File "/lib/python3.10/subprocess.py", line 966, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
File "/lib/python3.10/subprocess.py", line 1775, in _execute_child
    self.pid = _posixsubprocess.fork_exec(
```

I guess that's just a pyodide shortcoming?

**Reply by caleb-depotanalytics** [reply]
Or in retrospect, I guess that pyodide is functionally creating a linux container in the browser... Maybe it's possible to execute linux binaries from within the app? Would I just be wasting my time trying that out or should I give it a shot?

**Reply by whitphx** [reply]
> I guess that's just a pyodide shortcoming?

Yes.

I think neither Win or Linux binary works in the Pyodide environment. It provides a POSIX-like file system so the file paths look like Linux's ones, but it's not a Linux container.

**Reply by caleb-depotanalytics** [reply]
Okay, dang - thanks for all the answers on this!!
